src/schemapack/models.py

```diff
@@ -0,0 +1,413 @@
+    classes: dict[str, ClassDefinition] = Field(
+        ...,
+        description=(
+            "A mapping of class names to class definitions. Class names should use"
+            + " PascalCase."
+        ),
+        min_length=1,
+    )
+    root: Optional[str] = Field(
+        None,
+        description=(
+            "By default, schemapacks are unrooted meaning that they can be used to"
+            + " describe any number of instances of the classes contained in the"
+            + " schemapack."
+            + " Using this field, you may optionally specify the name of a class that"
+            + " acts as root. The schemapack will then be scoped to only"
+            + " describe a single instance of the root class along with its relations."
+        ),
+    )
+    self_path: Optional[Path] = Field(
+        None, description="The path to this schemapack.", exclude=True
+    )
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_schemapack_field_exists(cls, data: Any) -> Any:
+        """Checks that the schemapack field exists and points to a supported version
+        before doing anything else. Please note, this validation only takes place if the
+        data is passed as dict. However, the data can be pretty much anything.
+        """
+        if isinstance(data, dict):
+            if "schemapack" not in data:
+                raise PydanticCustomError(
+                    "MissingSchemaPackVersionError",
+                    (
+                        "Missing a `schemapack` field. Are you sure you have passed a"
+                        + " schemapack definition?"
+                    ),
+                )
+
+            if data["schemapack"] not in SUPPORTED_SCHEMA_PACK_VERSIONS:
+                raise PydanticCustomError(
+                    "UnsupportedSchemaPackVersionError",
+                    (
+                        "Unsupported schemapack version '{current_version}'."
+                        " Supported versions are: {supported_versions}"
+                    ),
+                    {
+                        "current_version": data["schemapack"],
+                        "supported_versions": SUPPORTED_SCHEMA_PACK_VERSIONS,
+                    },
+                )
+
+        return data
+
+    @field_validator("classes", mode="after")
+    @classmethod
+    def class_name_validator(
+        cls, v: dict[str, ClassDefinition]
+    ) -> dict[str, ClassDefinition]:
+        """Validates class names."""
+        invalid_relation_names = [
+            relation_name for relation_name in v if not relation_name.isidentifier()
+        ]
+
+        if invalid_relation_names:
+            raise PydanticCustomError(
+                "InvalidClassNameError",
+                (
+                    "Class names may only contain alphanumeric characters and"
+                    + " underscores. They must not start with a number."
+                    + " Got {number} invalid names: {relation_name}"
```

Comment:
```suggestion
                    + " Got {number} invalid names: {invalid_relation_names}"
```

Reply:
Or actually, should be called `invalid_class_names` here.

If you want to avoid duplication, you can also use Annotated Validators:
```
    classes: Annotated[dict[str, ClassDefinition], AfterValidator(check_identifiers)] = Field(...)
    relations: Annotated[dict[str, Relation], AfterValidator(check_identifiers)] = Field(...)
```

Reply:
Good suggestion. Here I was following Martin Fowlers rule of three. However, once the third repetition occurs I will do so.